x-pack/plugins/lens/public/shared_components/coloring/utils.ts

```diff
@@ -0,0 +1,279 @@
+) {
+  const { min: dataMin, max: dataMax } = getDataMinMax(params?.rangeType, dataBounds);
+  const minStopValue = params?.colorStops?.[0]?.stop ?? Infinity;
+  const maxStopValue = params?.colorStops?.[params.colorStops.length - 1]?.stop ?? -Infinity;
+  const overallMin = Math.min(dataMin, minStopValue);
+  const overallMax = Math.max(dataMax, maxStopValue);
+  return { min: overallMin, max: overallMax };
+}
+
+export function getDataMinMax(
+  rangeType: CustomPaletteParams['rangeType'] | undefined,
+  dataBounds: { min: number; max: number }
+) {
+  const dataMin = rangeType === 'number' ? dataBounds.min : DEFAULT_MIN_STOP;
+  const dataMax = rangeType === 'number' ? dataBounds.max : DEFAULT_MAX_STOP;
+  return { min: dataMin, max: dataMax };
+}
+
+/**
+ * This is a generic function to compute stops from the current parameters.
+ */
+export function getPaletteStops(
+  palettes: PaletteRegistry,
+  activePaletteParams: CustomPaletteParams,
+  // used to customize color resolution
+  {
+    prevPalette,
+    dataBounds,
+    mapFromMinValue,
+  }: { prevPalette?: string; dataBounds: { min: number; max: number }; mapFromMinValue?: boolean }
+) {
+  const { min: minValue, max: maxValue } = getOverallMinMax(activePaletteParams, dataBounds);
+  const interval = maxValue - minValue;
+  const { stops: currentStops, ...otherParams } = activePaletteParams || {};
+
+  if (activePaletteParams.name === 'custom' && activePaletteParams?.colorStops) {
+    // need to generate the palette from the existing controlStops
+    return shiftPalette(activePaletteParams.colorStops, maxValue);
+  }
+  // generate a palette from predefined ones and customize the domain
+  const colorStopsFromPredefined = palettes
+    .get(prevPalette || activePaletteParams?.name || defaultParams.name)
+    .getCategoricalColors(defaultParams.steps, otherParams);
+
+  const newStopsMin = mapFromMinValue ? minValue : interval / defaultParams.steps;
+
+  const stops = remapStopsByNewInterval(
+    colorStopsFromPredefined.map((color, index) => ({ color, stop: index })),
+    {
+      newInterval: interval,
+      oldInterval: colorStopsFromPredefined.length,
+      newMin: newStopsMin,
+      oldMin: 0,
+    }
+  );
+  return stops;
+}
+
+export function reversePalette(paletteColorRepresentation: ColorStop[] = []) {
+  const stops = paletteColorRepresentation.map(({ stop }) => stop);
+  return paletteColorRepresentation
+    .map(({ color }, i) => ({
+      color,
+      stop: stops[paletteColorRepresentation.length - i - 1],
+    }))
+    .reverse();
+}
+
+export function mergePaletteParams(
+  activePalette: PaletteOutput<CustomPaletteParams>,
+  newParams: CustomPaletteParams
+): PaletteOutput<CustomPaletteParams> {
+  return {
+    ...activePalette,
+    params: {
+      ...activePalette.params,
+      ...newParams,
+    },
+  };
+}
+
+function isValidPonyfill(colorString: string) {
+  // we're using an old version of chroma without the valid function
```

Comment:
Why do we need this? The version of chroma won't change, right?

Reply:
The newer version of `chroma` brings some breaking changes, so upgrade times I think will likely be not super shorts. This is a specific fix to validate hex values with alpha channel.
The alternative is to handle this by the EUI hex validation level, but from my understanding at the moment the hex4/hex8 format is likely not supported.